File: src/main/java/com/machfour/macros/objects/Serving.java
package com.machfour.macros.objects;

import com.machfour.macros.core.*;
import org.jetbrains.annotations.NotNull;

public class Serving extends MacrosEntityImpl<Serving> implements PortionMeasurement {

    private Food food;
    private QtyUnit qtyUnit;

    protected Serving(ColumnData<Serving> data, ObjectSource objectSource) {
        super(data, objectSource);
        food = null;
        qtyUnit = null;
    }
    @Override
    public Table<Serving> getTable() {
        return table();
    }
    public static Table<Serving> table() {
        return Schema.ServingTable.instance();
    }
    @Override
    public Factory<Serving> getFactory() {
        return factory();
    }
    public static Factory<Serving> factory() {
        return Serving::new;
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Serving && super.equals(o);
    }

    public Food getFood() {
        return food;
    }

    public void setFood(@NotNull Food f) {
        assert food == null && foreignKeyMatches(this, Schema.ServingTable.FOOD_ID, f);
        food = f;
    }

    @NotNull
    public String getQtyUnitAbbr() {
        return getData(Schema.ServingTable.QUANTITY_UNIT);
    }

    public QtyUnit getQtyUnit() {
        return qtyUnit;
    }

    public void setQtyUnit(@NotNull QtyUnit q) {
        assert getQtyUnit() == null && foreignKeyMatches(this, Schema.ServingTable.QUANTITY_UNIT, q);
        qtyUnit = q;
    }

    public long getFoodId() {
        return getData(Schema.ServingTable.FOOD_ID);
    }

    @NotNull
    @Override
    public String getName() {
        return getData(Schema.ServingTable.NAME);
    }

    public double getQuantity() {
        return getData(Schema.ServingTable.QUANTITY);
    }

    public boolean isDefault() {
        return getData(Schema.ServingTable.IS_DEFAULT);
    }

    // Measurement functions
    @Override
    public double unitMultiplier() {
        return getQuantity();
    }

    @Override
    public QtyUnit baseUnit() {
        return getQtyUnit();
    }

    @Override
    public boolean isVolumeMeasurement() {
        return getQtyUnit().isVolumeUnit();
    }

    @Override
    public boolean isServing() {
        return true;
    }
}

File: src/main/kotlin/com/machfour/macros/objects/Serving.kt
package com.machfour.macros.objects

import com.machfour.macros.core.*

class Serving internal constructor(data: ColumnData<Serving>, objectSource: ObjectSource)
    : MacrosEntityImpl<Serving>(data, objectSource), PortionMeasurement {

    companion object {
        // factory has to come before table if it's an instance variable
        val factory: Factory<Serving> = Factory { dataMap, objectSource -> Serving(dataMap, objectSource) }
        val table: Table<Serving>
            get() = Schema.ServingTable.instance
    }

    lateinit var food: Food
        private set

    val qtyUnitAbbr: String
        get() = getData(Schema.ServingTable.QUANTITY_UNIT)!!

    val qtyUnit: QtyUnit = QtyUnits.fromAbbreviation(qtyUnitAbbr)

    override val factory: Factory<Serving>
        get() = Companion.factory
    override val table: Table<Serving>
        get() = Companion.table

    val foodId: Long
        get() = getData(Schema.ServingTable.FOOD_ID)!!

    val quantity: Double
        get() = getData(Schema.ServingTable.QUANTITY)!!

    val isDefault: Boolean
        get() = getData(Schema.ServingTable.IS_DEFAULT)!!

    override fun equals(other: Any?): Boolean {
        return other is Serving && super.equals(other)
    }

    override fun hashCode(): Int {
        return super.hashCode()
    }

    fun initFood(f: Food) {
        assert(foreignKeyMatches(this, Schema.ServingTable.FOOD_ID, f))
        food = f
    }

    override val name: String = getData(Schema.ServingTable.NAME)!!

    // Measurement functions
    override val unitMultiplier = quantity

    override val baseUnit
        get() = qtyUnit

    override val isVolumeMeasurement
        get() = qtyUnit.isVolumeUnit

    override val isServing = true

}
